### ISSUE TITLE
Race condition when terminating a thread

### ISSUE THREAD

**Issue by jan11011977** [andymccurdy/redis-py#1134]
**Version**: Redis-py version 3.1.0, Redis: any version
**Platform**: Python 3.7 on Windows 10, but could happen on any platform
**Description**:

When a Gevent thread is terminated while talking to Redis, its connection will be recycled into the connection pool, even though the reply from Redis is still pending. The next thread to then use this connection will receive the pending reply from Redis.

Here is a code snippet that demonstrates the problem::

```import gevent
from gevent.monkey import patch_all
patch_all()
from redis import Redis


def mythread(redis):
    while True:
        redis.hget('test', 'key')


redis = Redis()
redis.hset('test', 'key', 'value')
redis.hset('test', 'otherkey', 'othervalue')

thread = gevent.spawn(mythread, redis)
gevent.sleep(0.1)
thread.kill()

result = redis.hget('test', 'otherkey')
print(result)```


The output I would expect: "othervalue"
The output I am seeing: "value"

This can cause some strange bugs, for example "invalid literal for int() with base 10" if you are running a HDEL command, but you are receiving the response of a HGET command.

Thanks!

**Comment by andymccurdy** [comment]
I believe this should already be fixed with code in the “healthy_connections” branch. Would you be able to test with the branch and let me know?

I still need to write a few more tests but I hope to merge that branch sometime this week.

**Comment by jan11011977** [comment]
Thanks Andy, yes I can confirm that it works correctly with the healthy_connections branch.

**Comment by andymccurdy** [comment]
Great. The healthy_connections branch was merged into the 3.2.0 release. I'll consider this closed then. Thanks!
